Restore support for GCC-4.8.

coryan opened the pull request:
With the latest version of nlohmann::json we can support GCC-4.8 again.

This fixes #798.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1731)
<!-- Reviewable:end -->

devbww commented on CMakeLists.txt:
```diff
@@ -27,14 +27,13 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 4.8)
         message(
             FATAL_ERROR
                 "GCC version must be at least 4.9. Older versions"
```
s/4.9/4.8/

coryan replied:
Fixed.

devbww commented on google/cloud/firestore/field_path.cc:
```diff
@@ -63,10 +61,33 @@ FieldPath FieldPath::Append(FieldPath const& field_path) const {
 }
 
 std::string FieldPath::ToApiRepr() const {
+#if defined(__GNUC__) && __GNUC__ == 4 && __GNUC_MINOR__ == 8
```
Do we like the "defined(\_\_GNUC\_\_)"?  If it wasn't defined, "\_\_GNUC\_\_ == 4" would be false anyway.  Or would that make this expression gratuitously different from other uses where that doesn't apply?

coryan replied:
There are a few other uses of `defined(__GNUC__)` in this file, so I would prefer to keep as-is for now. Later we can change all of them, I have no opinion on which style is "better".

Having said that, if we find ourselves testing for compiler names and versions in a lot of places we are doing something wrong.

devbww commented on google/cloud/firestore/field_path.cc:
```diff
@@ -63,10 +61,33 @@ FieldPath FieldPath::Append(FieldPath const& field_path) const {
+    return true;
+  };
+#else
+  auto is_simple = [](std::string const& part) {
```
"is_simple_field_name" perhaps?

coryan replied:
Fixed.  On second thought I removed the code using regex, seems like overkill anyway, happy to restore it if you think it is important.

devbww replied:
No ... I prefer the non-conditional too.

devbww commented on google/cloud/storage/status_or.h:
```diff
@@ -181,8 +185,10 @@ class StatusOr final {
   Status const&& status() const&& { return std::move(status_); }
-  //@}
+#endif  // GOOGLE_CLOUD_CPP_HAVE_CONST_REF_REF
+        //@}
```
Is the indentation change significant?  (And below.)

coryan replied:
No, clang-format just did that, we could insert a blank line or reorder the functions to make it look better.  Frankly I just let clang-format do its thing, most of the time it is good, and I rather worry about other things. I will change it if you think it is important, of course.

devbww replied:
I'm happy to let clang-format have its way (mostly) too.

coryan commented:
PTAL.

devbww commented:
It looks like there were some check errors, but I can't seem to see them.

coryan commented:
I fixed the permissions, you should be able to see the Kokoro builds now. I also fixed the build. For extra
fun it looks like we had a VM crash or BSOD in the middle of our build yay!
